### PR TITLE
Switch to lerna.js

### PR DESCRIPTION
**Pull request by silverbucket**
All repositories have been imported and everything uses yarn exclusively now. Lerna has been configured and I can run tests for all packages locally. Still some things to do:

- [x] Get Travis tests to work
- [x] Running Sockethub
- [x] Config
- [x] Managing PRs, versioning and releases. (done, sort of, needs to be tested)

The PR is using `independent` versioning for now, each time you run `lerna publish` you have to decide he release version of each package that has changes. We could manually enforce version unity for `sockethub*` packages, with the other packages independent. If it becomes a pain we can switch to using the same version for every package.

### CODE REVIEW

**Comment by skddc** [comment]
Just saw the request for review. Very excited about this getting into the final stage!

Could you expand on the open "config" to-do perhaps? No idea what that means without context. Thanks.

File: README.md [hunk 1]
```diff
@@ -2,133 +2,54 @@
...
-| Bitcoin (e.g. sending signed transactions to a node) | *(to be evaluated)* <kbd>send</kbd> <kbd>receive</kbd> ... |
-| Ethereum (e.g. contract events) | *(to be evaluated)* |
-| WebRTC (signaling) | *(to be evaluated)* |
-| Signal | *(to be evaluated)* |
-| [MicroPub](https://indieweb.org/micropub) (IndieWeb, ...) | *(to be evaluated)* |
-| [OStatus](https://en.wikipedia.org/wiki/OStatus) (GNUSocial, Mastodon, ...) | *(to be evaluated)* |
+### Dependencies
 
-# Setup
+```$ yarn run install```
```
Comment (raucao) [inline README.md]:
```suggestion
```$ yarn install```
```

File: README.md [hunk 2]
```diff
@@ -2,133 +2,54 @@
...
 
-`$ yarn install`
+### Tests
 
-# Running
+```$ yarn run test```
```
Comment (raucao) [inline README.md]:
```suggestion
```$ yarn test```
```

**Comment by galfert** [comment]
Shouldn't Lerna be a `devDependency` in the `package.json`? Otherwise the readme is missing that Lerna needs to be installed globally.

File: README.md [hunk 3]
```diff
@@ -2,133 +2,54 @@
...
 
-For development purposes, with examples enabled, run:
+### Run
 
-`$ DEBUG=sockethub* yarn run dev`
+```$ yarn run start```
```
Comment (raucao) [inline README.md]: Shouldn't this rather document how to run Sockethub from a normal npm-installed module's executable? Running directly from a source repo is something you'd usually do for development, but not production.

File: README.md [hunk 4]
```diff
@@ -2,133 +2,54 @@
...
-  --port       : port to bind to
-```
+* [packages/activity-streams.js](packages/activity-streams.js)
+* [packages/irc2as](packages/irc2as)
+* [packages/sockethub](packages/sockethub)
+* [packages/sockethub-platform-dummy](packages/sockethub-platform-dummy)
+* [packages/sockethub-platform-feeds](packages/sockethub-platform-feeds)
+* [packages/sockethub-platform-irc](packages/sockethub-platform-irc)
+* [packages/sockethub-platform-xmpp](packages/sockethub-platform-xmpp)
+* [packages/sockethub-schemas](packages/sockethub-schemas)
```
Comment (raucao) [inline README.md]: I would remove the "packages/" prefixes here, so that the list shows the pure package names as links.

File: README.md [hunk 5]
```diff
@@ -2,133 +2,54 @@
...
 
 # License
```
Comment (raucao) [inline README.md]: The other headings have been moved to level 2 (which I think is much better), so License and Credits should be changed to level 2 as well.

File: README.md [hunk 6]
```diff
@@ -2,133 +2,54 @@
...
-| [Feeds](https://github.com/sockethub/sockethub-platform-feeds) *(RSS, Atom)* | <kbd>![completed](http://sockethub.org/res/img/checkmark.png) fetch</kbd> |
-| [IRC](https://github.com/sockethub/sockethub-platform-irc) | <kbd>![completed](http://sockethub.org/res/img/checkmark.png) send</kbd> <kbd>![completed](http://sockethub.org/res/img/checkmark.png) join</kbd> <kbd>![completed](http://sockethub.org/res/img/checkmark.png) leave</kbd> <kbd>![completed](http://sockethub.org/res/img/checkmark.png) observe</kbd> <kbd>![completed](http://sockethub.org/res/img/checkmark.png) update</kbd> |
-
-## Out of date
-
-Implemented in a previous major release of Sockethub, never ported to the current release and unmaintained.
+**For more information on Sockethub see [Sockethub README](packages/sockethub/README.md)**
```
Comment (raucao) [inline README.md]:
```suggestion
**For more information on Sockethub see the [Sockethub README](packages/sockethub/README.md) and the [Sockethub Wiki](https://github.com/sockethub/sockethub/wiki)**.
```

**Comment by silverbucket** [comment]
> Could you expand on the open "config" to-do perhaps? No idea what that means without context. Thanks.

This was figuring out a way to have config apply when running sockethub from the repo root, whereas it was previously designed to be run from the sockethub root. Got it working but not sure if it's ideal (see repo root package.json for the start and dev commands)

**Comment by skddc** [comment]
What happened to the README? It looks like it was deleted from the root dir, but that's where it's needed for developers and people looking at the repo.

**Comment by silverbucket** [comment]
> What happened to the README that I reviewed? It looks like it was deleted from the root dir, but that's where it's needed for developers and other people looking at the repo. Now it's also not possible anymore to easily commit my suggestions.

It's symlinked in the root from the `packages/sockethub/` directory.

**Comment by skddc** [comment]
> It's symlinked in the root from the `packages/sockethub/` directory.

Yes, but there was another README in the package dir before, which was linked in the shared README. It's rather difficult to follow as someone else, because the changeset is so huge, and also because all my comments are now not linked to the document anymore.

**Comment by silverbucket** [comment]
> > It's symlinked in the root from the `packages/sockethub/` directory.
> 
> Yes, but there was another README in the package dir before, which was linked in the shared README. It's rather difficult to follow as someone else, because the changeset is so huge, and also because all my comments are now not linked to the document anymore.

That file was there before I marked the repo as ready for review (it was still in draft) I found I was splitting the same or similar info across the two readme files so decided to consolidate. Not sure if it's the best idea though going forward. I am just trying to avoid doc duplication.

**Comment by skddc** [comment]
You had requested our reviews, which is why I added the comments. But also, there doesn't seem to be any consolidation, because the new package README seems to be the same as the previous shared README, and I cannot easily see what happened to the Sockethub package README. As I said, it's just very difficult to follow, and it would also be useful if I didn't have to add all my suggestions and review comments again. I spent the time already.

**Comment by silverbucket** [comment]
>  But also, there doesn't seem to be any consolidation, because the new package README seems to be the same as the previous shared README, 

Which is why I consolidated them, they were mostly duplicates. 

> and I cannot easily see what happened to the Sockethub package README. 

The Sockethub README is the only README, the root of the repo is a symlink to the Sockethub README (in the sockethub package). 

> As I said, it's just very difficult to follow, and it would also be useful if I didn't have to add all my suggestions and review comments again. I spent the time already.

I see your 6 comments even though they are flagged "outdated". If there are more, that are missing let me know. 

> So yeah, please don't request reviews, if stuff isn't ready to review! 
> We're all investing our own time here, so it's important to not waste it. 

I marked the PR as ready for review after I removed the file. I added your names to reviewers when the PR was still flagged as a draft, and in the description not all of the checkmarks checked. No ones time was wasted, your feedback is still there. 

More importantly, I don't even know if it makes sense to link the READMEs (so if you have thoughts on that I'm open to them) but it seemed like the best way to avoid duplicate info at the time. Your comments about module usage vs repo usage has me wondering how to best satisfy both cases. 

> And also, please use the review comments and suggestions that I already added, in whichever way that's easiest for you now. Thanks!

Of course! Shouldn't be a problem, thanks for the feedback!

**Comment by skddc** [comment]
> More importantly, I don't even know if it makes sense to link the READMEs (so if you have thoughts on that I'm open to them) but it seemed like the best way to avoid duplicate info at the time. Your comments about module usage vs repo usage has me wondering how to best satisfy both cases.

I think the README in the repo root should give a quick intro of what Sockethub is, and what one can find in this repo, while the package READMEs should contain the information that is actually relevant to the specific package. Also, I think one should avoid symlinks like the plague in Git repos in general, because they can break both on the Web as well as with non-POSIX-compliant operating systems.

**Comment by skddc** [comment]
> Your comments about module usage vs repo usage has me wondering how to best satisfy both cases

I thought the entire raison d'être of Lerna is that you can have all source code in a single Git repo, but publish all modules separately on npm. The repo should only be "used" for development in my opinion, while production usage should default to versioned, released packages.

**Comment by galfert** [comment]
I was able to run the tests and Sockethub locally.

Have you seen [my question from yesterday](https://github.com/sockethub/sockethub/pull/250#issuecomment-581502624)?

> Shouldn't Lerna be a `devDependency` in the `package.json`? Otherwise the readme is missing that Lerna needs to be installed globally.

**Comment by silverbucket** [comment]
> I was able to run the tests and Sockethub locally.
> 
> Have you seen [my question from yesterday](https://github.com/sockethub/sockethub/pull/250#issuecomment-581502624)?
> 
> > Shouldn't Lerna be a `devDependency` in the `package.json`? Otherwise the readme is missing that Lerna needs to be installed globally.

Yup I saw it, just added the devDependency

**Comment by silverbucket** [comment]
@galfert @skddc updated based on both of your feedback, made a README just for the monorepo root, and then sockethub specifics in that package dir. Ready for review / approval if no other issues found.

File: README.md [hunk 7]
```diff
@@ -36,6 +36,8 @@ For production, with examples disabled.
 
 `$ DEBUG=sockethub* yarn run start`
 
+*For more infor on configuration options, see the [Sockethub README](packages/sockethub/README.md#environment-variables)* section on evironment variables.*
```
Comment (raucao) [inline README.md]:
```suggestion
*For more info on configuration options, see the [Sockethub README](packages/sockethub/README.md#environment-variables)* section on evironment variables.*
```

**Comment by skddc** [comment]
Yay! :clap: 

Want to release a `3.2.0-rc.1` pre-release from this, so we can test an actual release?

**Comment by silverbucket** [comment]
Assuming everything went well, we should have the following pre-releases up on npm: - activity-streams: 4.3.0 => 4.3.1-alpha.0
 - irc2as: 3.0.0 => 3.0.1-alpha.0
 - sockethub-platform-dummy: 2.0.0 => 2.0.1-alpha.0
 - sockethub-platform-feeds: 2.0.1 => 2.0.2-alpha.0
 - sockethub-platform-irc: 3.4.1 => 3.4.2-alpha.0
 - sockethub-platform-xmpp: 3.2.0 => 3.2.1-alpha.0
 - sockethub-schemas: 2.0.0 => 2.0.1-alpha.0
 - sockethub: 3.1.2 => 3.1.3-alpha.0

**Comment by silverbucket** [comment]
Got that sorted, and published `sockethub-3.1.3-alpha.1` but still don't have access to the sockethub command. Hmm...

**Comment by skddc** [comment]
> but still don't have access to the sockethub command

Shouldn't the `main` point to `dist/index.js`? Also, there's no `bin` defined in that `package.json`. The `bin` should also point to the JS file to execute, and is then installed to `/usr/bin/sockethub`.

**Comment by silverbucket** [comment]
@skddc no the executable is `bin/sockethub` not dist/index... and yeah - it should be installing to a bin dir as you describe, but it's not... I'm not sure why. Check out the latest release (fixed some production vs dev checks) `sockethub@3.1.3-alpha.2`

If I run the executable directly:
```
$ /usr/local/lib/node_modules/sockethub/bin/sockethub

UNCAUGHT EXCEPTION

Error: Cannot find module 'irc-socket'
Require stack:
- /usr/local/lib/node_modules/sockethub/node_modules/sockethub-platform-irc/index.js
- /usr/local/lib/node_modules/sockethub/dist/bootstrap/platforms.js
- /usr/local/lib/node_modules/sockethub/dist/bootstrap/init.js
- /usr/local/lib/node_modules/sockethub/dist/sockethub.js
- /usr/local/lib/node_modules/sockethub/bin/sockethub
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:625:15)
    at Function.Module._load (internal/modules/cjs/loader.js:527:27)
    at Module.require (internal/modules/cjs/loader.js:683:19)
    at require (internal/modules/cjs/helpers.js:16:16)
    at Object.<anonymous> (/usr/local/lib/node_modules/sockethub/node_modules/sockethub-platform-irc/index.js:26:15)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Module.load (internal/modules/cjs/loader.js:643:32)
    at Function.Module._load (internal/modules/cjs/loader.js:556:12)
    at Module.require (internal/modules/cjs/loader.js:683:19)
    at require (internal/modules/cjs/helpers.js:16:16)
    at Object.platformLoad [as default] (/usr/local/lib/node_modules/sockethub/dist/bootstrap/platforms.js:50:27)
    at Object.<anonymous> (/usr/local/lib/node_modules/sockethub/dist/bootstrap/init.js:12:38)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Module.load (internal/modules/cjs/loader.js:643:32)

```

**Comment by skddc** [comment]
> it should be installing to a bin dir as you describe, but it's not...

See my edited comment. There's no `bin` property defined, so nothing is telling npm to install an executable.

**Comment by silverbucket** [comment]
@skddc OK cool - it installs and we have access to the sockethub executable. However we're still having module dependency issues.

```
$ sockethub --help

UNCAUGHT EXCEPTION

Error: Cannot find module 'irc-socket'
Require stack:
- /usr/local/lib/node_modules/sockethub/node_modules/sockethub-platform-irc/index.js
- /usr/local/lib/node_modules/sockethub/dist/bootstrap/platforms.js
- /usr/local/lib/node_modules/sockethub/dist/bootstrap/init.js
- /usr/local/lib/node_modules/sockethub/dist/sockethub.js
- /usr/local/lib/node_modules/sockethub/bin/sockethub
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:625:15)
    at Function.Module._load (internal/modules/cjs/loader.js:527:27)
    at Module.require (internal/modules/cjs/loader.js:683:19)
    at require (internal/modules/cjs/helpers.js:16:16)
    at Object.<anonymous> (/usr/local/lib/node_modules/sockethub/node_modules/sockethub-platform-irc/index.js:26:15)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Module.load (internal/modules/cjs/loader.js:643:32)
    at Function.Module._load (internal/modules/cjs/loader.js:556:12)
    at Module.require (internal/modules/cjs/loader.js:683:19)
    at require (internal/modules/cjs/helpers.js:16:16)
    at Object.platformLoad [as default] (/usr/local/lib/node_modules/sockethub/dist/bootstrap/platforms.js:50:27)
    at Object.<anonymous> (/usr/local/lib/node_modules/sockethub/dist/bootstrap/init.js:12:38)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Module.load (internal/modules/cjs/loader.js:643:32)

```

**Comment by silverbucket** [comment]
Ah, looks like dependency vs devDependency issues. Going to go through all the package.json files and double check these. Then make another release.

**Comment by silverbucket** [comment]
Alright! sockethub is now published and seems to work. The examples look broken due to some path issue, I can't look at it now, but if someone is able to track this down, awesome. Otherwise I'll keep at this over the weekend in small chunks.

File: packages/sockethub/package.json [hunk 8]
```diff
@@ -5,7 +5,8 @@
   "private": false,
   "author": "Nick Jennings <nick@silverbucket.net>",
   "license": "LGPL-3.0+",
-  "main": "./bin/sockethub",
+  "main": "yarn run utart",
```
Comment (raucao) [inline packages/sockethub/package.json]: I don't think that's what "main" is for. It usually points to a JS file to use with programs like nodemon for example.

Comment (galfert) [reply packages/sockethub/package.json]: The `main` property should point to the file that contains the export you want to be given when doing a `require('sockethub')`. See https://docs.npmjs.com/files/package.json#main

Comment (silverbucket) [reply packages/sockethub/package.json]: Right, makes sense now :) thanks - I knew this but somewhere along the line conflated it with the bin property. Anyway, it's fixed now.

**Comment by silverbucket** [comment]
Alright! managed to work out the kinds, please try it out yourselves and try to break it. It's at alpha7 now (`sockethub@3.1.3-alpha.7`). I gotta say, really loving the ease of management with lerna so far. Working out these issues and publishing has been a total breeze. 

One thing I'd like to look into is to find a way to avoid publishing before I've build the dist files (compiled from TS to JS). Not sure how best to do this, if anyone has any suggestions.

**Comment by skddc** [comment]
> One thing I'd like to look into is to find a way to safeguard against publishing before I've build the dist files (compiled from TS to JS). Not sure how best to do this, if anyone has any suggestions.

The best way to do this is not to safeguard, but to build in an npm script/hook, so that the dist files are also part of the publishing/version commit. With npm, we usually do that in the `version` hook, which is executed after increasing the version number, but before committing the changes. For example:

https://github.com/67P/hyperchannel/blob/master/package.json#L29

I would guess that this is possible somehow with lerna as well.

**Comment by silverbucket** [comment]
> > One thing I'd like to look into is to find a way to safeguard against publishing before I've build the dist files (compiled from TS to JS). Not sure how best to do this, if anyone has any suggestions.
> 
> I think the best way to do this is not to safeguard, but to build in an npm script/hook, so that the dist files are also part of the publishing/version commit. With npm, we usually do that in the `version` hook, which is executed after increasing the version number, but before committing the changes. For example:
> 
> https://github.com/67P/hyperchannel/blob/master/package.json#L29
> 
> I would guess that this is possible somehow with lerna as well.

Right, I was thinking of something like this. Making a `release` command in the monorepo-root package.json, which will run `lerna run build && lerna release` however there also needs to be a commit in there before lerna release is called, and I'm not sure if it's smart to have a commit built in to a command like that. eg. `lerna run build && git commit -m "build dist files for release" && lerna release`

**Comment by skddc** [comment]
That's not what I meant. What you describe would still be separate commits between the build and release. I mean that there should be some kind of release hook that you can use with lerna, same as with every [npm script](https://docs.npmjs.com/misc/scripts).

**Comment by silverbucket** [comment]
> That's not what I meant. What you describe would still be separate commits between the build and release. I mean that there should be some kind of release hook that you can use with lerna, same as with every [npm script](https://docs.npmjs.com/misc/scripts).

I see what you mean, will try to figure out an approach here.

**Comment by skddc** [comment]
Found it: https://github.com/lerna/lerna/tree/master/commands/version#lifecycle-scripts

**Comment by galfert** [comment]
I was able to install Sockethub globally from NPM and run it without problems.

But when I install it locally as a dependency of another project and try to run it via

```bash
DEBUG=sockethub* ./node_modules/sockethub/bin/sockethub
```

I get the following error:

```bash
  sockethub:bootstrap:config initializing config +0ms
  sockethub:schemas loading sockethub activity stream schemas +0ms
  sockethub:bootstrap:platforms loading platforms +0ms
  sockethub:bootstrap:init running init routines +0ms
  sockethub:bootstrap:platforms finding and registering platforms from package list +1ms
  sockethub:bootstrap:platforms registering dummy platform +0ms

UNCAUGHT EXCEPTION

Error: Cannot find module './../../node_modules/sockethub-platform-dummy/package.json'
Require stack:
- /Users/galfert/code/playground/sockethub-install-test/node_modules/sockethub/dist/bootstrap/platforms.js
- /Users/galfert/code/playground/sockethub-install-test/node_modules/sockethub/dist/bootstrap/init.js
- /Users/galfert/code/playground/sockethub-install-test/node_modules/sockethub/dist/sockethub.js
- /Users/galfert/code/playground/sockethub-install-test/node_modules/sockethub/bin/sockethub
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:797:15)
    at Function.Module._load (internal/modules/cjs/loader.js:690:27)
    at Module.require (internal/modules/cjs/loader.js:852:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.platformLoad [as default] (/Users/galfert/code/playground/sockethub-install-test/node_modules/sockethub/dist/bootstrap/platforms.js:52:37)
    at Object.<anonymous> (/Users/galfert/code/playground/sockethub-install-test/node_modules/sockethub/dist/bootstrap/init.js:12:38)
    at Module._compile (internal/modules/cjs/loader.js:959:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:995:10)
    at Module.load (internal/modules/cjs/loader.js:815:32)
    at Function.Module._load (internal/modules/cjs/loader.js:727:14)
```

**Comment by silverbucket** [comment]
> I was able to install Sockethub globally from NPM and run it without problems.
> 
> But when I install it locally as a dependency of another project and try to run it via
> 
> ```shell
> DEBUG=sockethub* ./node_modules/sockethub/bin/sockethub
> ```
> 
> I get the following error:
> 
> ```shell
>   sockethub:bootstrap:config initializing config +0ms
>   sockethub:schemas loading sockethub activity stream schemas +0ms
>   sockethub:bootstrap:platforms loading platforms +0ms
>   sockethub:bootstrap:init running init routines +0ms
>   sockethub:bootstrap:platforms finding and registering platforms from package list +1ms
>   sockethub:bootstrap:platforms registering dummy platform +0ms
> 
> UNCAUGHT EXCEPTION
> 
> Error: Cannot find module './../../node_modules/sockethub-platform-dummy/package.json'
> Require stack:
> - /Users/galfert/code/playground/sockethub-install-test/node_modules/sockethub/dist/bootstrap/platforms.js
> - /Users/galfert/code/playground/sockethub-install-test/node_modules/sockethub/dist/bootstrap/init.js
> - /Users/galfert/code/playground/sockethub-install-test/node_modules/sockethub/dist/sockethub.js
> - /Users/galfert/code/playground/sockethub-install-test/node_modules/sockethub/bin/sockethub
>     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:797:15)
>     at Function.Module._load (internal/modules/cjs/loader.js:690:27)
>     at Module.require (internal/modules/cjs/loader.js:852:19)
>     at require (internal/modules/cjs/helpers.js:74:18)
>     at Object.platformLoad [as default] (/Users/galfert/code/playground/sockethub-install-test/node_modules/sockethub/dist/bootstrap/platforms.js:52:37)
>     at Object.<anonymous> (/Users/galfert/code/playground/sockethub-install-test/node_modules/sockethub/dist/bootstrap/init.js:12:38)
>     at Module._compile (internal/modules/cjs/loader.js:959:30)
>     at Object.Module._extensions..js (internal/modules/cjs/loader.js:995:10)
>     at Module.load (internal/modules/cjs/loader.js:815:32)
>     at Function.Module._load (internal/modules/cjs/loader.js:727:14)
> ```

@galfert That is an unsupported way to run sockethub, use only documented ways or if you see this in the docs somewhere point it out so we can correct it. Should be `yarn start`

**Comment by silverbucket** [comment]
> > I was able to install Sockethub globally from NPM and run it without problems.
> > But when I install it locally as a dependency of another project and try to run it via
> > ```shell
> > DEBUG=sockethub* ./node_modules/sockethub/bin/sockethub
> > ```
> > 
> > 
> > I get the following error:

> 
> @galfert That is an unsupported way to run sockethub, use only documented ways or if you see this in the docs somewhere point it out so we can correct it. Should be `yarn start`

Oops, sorry misunderstood. Though, why would you install it as a local dependency, and then run the binary? It should work though, will look into this.

File: package.json [hunk 9]
```diff
@@ -7,6 +7,7 @@
     "yarn": "^1.21.1"
   },
   "scripts": {
+    "prepublish": "lerna run build && gid add **/dist/*",
```
Comment (raucao) [inline package.json]:
```suggestion
    "prepublish": "lerna run build && git add **/dist/*",
```

Comment (raucao) [reply package.json]: I think what you want is the `version` hook though. And not `pre`, because that's before the version is incremented.

**Comment by galfert** [comment]
> Oops, sorry misunderstood. Though, why would you install it as a local dependency, and then run the binary? It should work though, will look into this.

One reason is that your project might only work with a certain version of Sockethub, so you want to pin the version as a `devDepency`. E.g. in Hyperchannel the current `master` only works with Sockethub v3.0.1, while the branch `feature/sockethub_master_compatibility` works with Sockethub v3.1.2. So having those specific versions specified in the `package.json` would help a lot to run the correct SH while switching branches and developing.

**Comment by silverbucket** [comment]
Alright, as they say, 13th time's a charm! It looks like I ironed out all the kinks in regards to path finding for the platform modules (as @galfert reported) and having a version hook which builds any package which a `build` directive, and checks in the resulting files as `Publish`.  Can everyone try to break `sockethub@3.1.3-alpha.13` ?

**Comment by skddc** [comment]
Very cool!

I still see a bunch of warnings on npm install:

```
npm WARN saveError ENOENT: no such file or directory, open '/home/basti/src/playground/sockethub/package.json'
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN enoent ENOENT: no such file or directory, open '/home/basti/src/playground/sockethub/package.json'
npm WARN ws@7.2.1 requires a peer of bufferutil@^4.0.1 but none is installed. You must install peer dependencies yourself.
npm WARN ws@7.2.1 requires a peer of utf-8-validate@^5.0.2 but none is installed. You must install peer dependencies yourself.
npm WARN sockethub No description
npm WARN sockethub No repository field.
npm WARN sockethub No README data
npm WARN sockethub No license field.
```

**Comment by skddc** [comment]
I was able to run sockethub from an installed package, but running `yarn start` in the source repo results in a very weird exception, which includes my npm user config for some reason:

```
Trace: 
  TypeError: Cannot create property 'username' on string '{"email":"sebastian@kip.pe","username":"skddc","always-auth":""}'
      at set (/home/basti/src/sockethub/sockethub/node_modules/yarn/lib/cli.js:53901:28)
      at set (/home/basti/src/sockethub/sockethub/node_modules/yarn/lib/cli.js:53915:14)
      at set (/home/basti/src/sockethub/sockethub/node_modules/yarn/lib/cli.js:53895:16)
      at Function.objectPath.set (/home/basti/src/sockethub/sockethub/node_modules/yarn/lib/cli.js:53948:14)
      at NpmRegistry.mergeEnv (/home/basti/src/sockethub/sockethub/node_modules/yarn/lib/cli.js:96018:18)
      at NpmRegistry.<anonymous> (/home/basti/src/sockethub/sockethub/node_modules/yarn/lib/cli.js:31899:12)
      at Generator.next (<anonymous>)
      at step (/home/basti/src/sockethub/sockethub/node_modules/yarn/lib/cli.js:310:30)
      at /home/basti/src/sockethub/sockethub/node_modules/yarn/lib/cli.js:328:14
      at new Promise (<anonymous>)
```

**Comment by skddc** [comment]
OK, so `yarn run dev` works, but `yarn start` fails with the above error.

**Comment by silverbucket** [comment]
> OK, so `yarn run dev` works, but `yarn start` fails with the above error.
> 
> Also, `yarn run dev` is extremely slow for some reason. Takes a long time for all launch steps to complete.

That's due to using `ts-node` as opposed to pre-compiling ts to js

**Comment by silverbucket** [comment]
@skddc I'm not able to recreate the behavior you are describing with `yarn start` it runs fine for me. I'm running `1.21.1` which was also the base version described in the `package.json` so you seem to be running a very old one.

**Comment by gregkare** [comment]
I've been able to run sockethub from the latest alpha npm package, as well as from the repo. Everything works great here

File: README.md [hunk 10]
```diff
@@ -84,57 +36,23 @@ For production, with examples disabled.
...
-
-# Environment Variables
-
-* PORT
-Defaults to `10550`
-* HOST
-Defaults to `localhost`
-* DEBUG
-Specify the namespace to console log, ie. `sockethub*` will print all sockethub related debug statements, whereas `*` will also print any other modules debug statements that use the `debug` module.
+*For more info on configuration options, see the [Sockethub README](packages/sockethub/README.md#environment-variables)* section on evironment variables.*
```
Comment (gregkare) [inline README.md]: Typo, `evironment` instead of `environment`

**Comment by skddc** [comment]
> @skddc I'm not able to recreate the behavior you are describing with `yarn start` it runs fine for me. I'm running `1.21.1` which was also the base version described in the `package.json` so you seem to be running a very old one.

I'm running 1.22.0.

**Comment by skddc** [comment]
Here's the detailed yarn error log:

```
Arguments: 
   2   /home/basti/.nvm/versions/node/v12.14.0/bin/node /home/basti/.nvm/versions/node/v12.14.0/bin/yarn run build
   3 
   4 PATH: 
   5   /tmp/yarn--1581351114517-0.12419144134585491:/home/basti/src/sockethub/sockethub/packages/sockethub/node_modules/.bin:/home/basti/.config/yarn/link/node_modules/.bin:
   6 
   7 Yarn version: 
   8   1.22.0
   9 
  10 Node version: 
  11   12.14.0
  12 
  13 Platform: 
  14   linux x64
  15 
  16 Trace: 
  17   TypeError: Cannot create property 'username' on string '{"email":"sebastian@kip.pe","username":"skddc","always-auth":""}'
  18       at set (/home/basti/.nvm/versions/node/v12.14.0/lib/node_modules/yarn/lib/cli.js:53901:28)
  19       at set (/home/basti/.nvm/versions/node/v12.14.0/lib/node_modules/yarn/lib/cli.js:53915:14)
  20       at set (/home/basti/.nvm/versions/node/v12.14.0/lib/node_modules/yarn/lib/cli.js:53895:16)
  21       at Function.objectPath.set (/home/basti/.nvm/versions/node/v12.14.0/lib/node_modules/yarn/lib/cli.js:53948:14)
  22       at NpmRegistry.mergeEnv (/home/basti/.nvm/versions/node/v12.14.0/lib/node_modules/yarn/lib/cli.js:96079:18)
  23       at NpmRegistry.<anonymous> (/home/basti/.nvm/versions/node/v12.14.0/lib/node_modules/yarn/lib/cli.js:31899:12)
  24       at Generator.next (<anonymous>)
  25       at step (/home/basti/.nvm/versions/node/v12.14.0/lib/node_modules/yarn/lib/cli.js:310:30)
  26       at /home/basti/.nvm/versions/node/v12.14.0/lib/node_modules/yarn/lib/cli.js:328:14
  27       at new Promise (<anonymous>)
```

**Comment by gregkare** [comment]
Looks like something from your `~/.npmrc` that yarn does not like

**Comment by skddc** [comment]
It's a very vanilla npmrc:

```
registry.npmjs.org/:username=skddc
registry.npmjs.org/:email=sebastian@kip.pe
registry.npmjs.org/:always-auth=false
registry.npmjs.org/:_password="123456789="
registry.npmjs.org/:_authToken=123456-123456-123456
progress=true
sign-git-tag=true
python=/usr/bin/python2
```

**Comment by gregkare** [comment]
On my system when I use this config file it's the `always-auth` line that's raising an exception

**Comment by skddc** [comment]
That doesn't change anything for me, unfortunately. But [`always-auth` defaults to `false`](https://docs.npmjs.com/misc/config#always-auth) anyway, so I removed it permanently nonetheless.

**Comment by skddc** [comment]
I found a bunch of issues and "solutions" in the yarn repo, some of which say it was a malformed `.npmrc` (which is clearly not the case for me), while others say it was actually a malformed `package.json` somewhere up the directory tree.

None of the issues have actually been closed, so it's still a problem with Yarn in general. :/

**Comment by skddc** [comment]
By the way, I was wondering, @galfert: what good is it to have `yarn` in the dev dependencies, when you have to have `yarn` to install the dev dependencies in the first place?

Also, apparently installing Yarn from npm has been deprecated altogether at some point, and now they have instructions for a wide variety of OSs and distros: https://classic.yarnpkg.com/en/docs/install#arch-stable

**Comment by skddc** [comment]
Just to confirm: I removed `yarn` from my global npm packages as well as the sockethub repo's package config and lockfile, and tried again with a vanilla yarn install from the Arch repo. The issue is still the same as before, so that bit seems irrelevant.

**Comment by galfert** [comment]
> By the way, I was wondering, @galfert: what good is it to have `yarn` in the dev dependencies, when you have to have `yarn` to install the dev dependencies in the first place?

Good point. I think I wrote that when installing Sockethub has a dependency of another project. But it doesn't really make sense.

**Comment by silverbucket** [comment]
I can't replicate the issue @skddc is having at all. Strange, not sure what else to try. To confirm, neither @galfert nor @gregkare can reproduce as well?

**Comment by skddc** [comment]
"NPM gives you trouble? Just use Yarn." Now you have Yarn trouble. :sneezing_face:

**Comment by silverbucket** [comment]
@skddc and if you delete/move the ~/.npmrc file does it work? And try adding one line at a time back?

Mine:
```
$ cat ~/.npmrc
//registry.npmjs.org/:_authToken=cbc5a5c1-290e-4292-add2-2663caedb574
```

**Comment by skddc** [comment]
Yes, then it works. As soon as I remove the valid username property for the npm registry, it doesn't fail anymore. But that config couldn't be any more valid if it tried.

**Comment by skddc** [comment]
By the way, I think you forgot to check in the yarn lockfiles in one of the recent dependency update commits. I have changes in all of them when I yarn-install from the latest one.

**Comment by silverbucket** [comment]
Also having a `yarn` issue with `jsdoc2md` (https://github.com/jsdoc2md/jsdoc-to-markdown/issues/206) when I did an `npm install` and `yarn install`, grabbing all packages in node_modules with versions, I got 8 differences. Some where missing entirely while others were major versions off. Weird.

**Comment by gregkare** [comment]
> I can't replicate the issue @skddc is having at all. Strange, not sure what else to try. To confirm, neither @galfert nor @gregkare can reproduce as well?

No, I can't reproduce this yarn issue. As I said above I tried using the exact same content as @skddc, for me removing the `always-auth` line (that defaults to false anyway) works, I can run `yarn start`.

I was using yarn 1.22.0 too, installed from npm

I've just switched to installing 1.22.0 from Homebrew and still not getting the issue

**Comment by silverbucket** [comment]
@gregkare i pinged you about the change request, it still says you have one request. Can you mark it ok? I fixed the typo you pointed out.

**Comment by gregkare** [comment]
done

**Pull request merged**